Documentation/building/code-coverage.md

```diff
@@ -56,16 +56,16 @@ This will do the build and testing as with the normal ```build```, but it will r
 
 You can also build and test with code coverage for a particular test project rather than for the whole repo.  Normally to build and test a particular test suite, from the same directory as that test suite's .csproj, you'd run:
 
-    msbuild /t:BuildAndTest
+    dotnet msbuild /t:BuildAndTest
```

Comment:
The plan is to keep both `msbuild` and `dotnet msbuild` working so I wonder if we should simply make a note of that somewhere and just reference `msbuild` in the instructions.

Reply:
Question, why should we continue advertising the Desktop msbuild if using the dotnet msbuild is (finally) possible in corefx (from a feature perspective). I don't see any compelling reason to stick with the Desktop flavor. We now already require the .NET Core SDK to be installed which means in 99% of the cases dotnet will be in the path.

Reply:
That is a fair question. Last I checked even if you called `dotnet msbuild` on a windows machine with full msbuild it would use the full version so I think we are going to need to keep our builds working for both and in such a case we should allow folks to use either.  In the past I know the performance was much better with full msbuild over core msbuild but I'm not sure if that is still the case or not, but another potential reason.

Reply:
While I thought dotnet msbuild would use the full msbuild your correct that doesn't seem to be the case from some quick testing. However the full msbuild is still used when building from VS, so we will need to maintain our build infrastructure to work with both. So it is really just a matter of which do we advertise in our docs. I'm OK advertising the `dotnet msbuild` and just adding a note that says you can also call msbuild on our projects.

Reply:
I'll probably keep using desktop just because it's quicker to type.

Reply:
From talking with @nguerrera, the current recommendation has been to put `dotnet.exe` on the `%PATH%`, and VS/MSBuild will be able to find the SDK using the resolution logic.  

So this means that you can either install the .NET Core SDK that we require (for now, the 3.0.100-preview version until 2.1.3 builds start coming out) into `Program Files`, which will then be on the PATH.  
Or you can set your PATH to include `corefx/Tools/dotnetcli`, which will then resolve the SDK from the Tools directory.
Or I guess you could install that version anywhere on your machine, and put it on the PATH.

I tested this in a normal command prompt:

```
F:\DotNetTest\TestConsole>"C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\MSBuild.exe" /pp:output.pp

F:\DotNetTest\TestConsole>set PATH=F:\corefx\Tools\dotnetcli\;%PATH%

F:\DotNetTest\TestConsole>"C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\MSBuild.exe" /pp:output2.pp
```

`output.pp` contains:  `C:\Program Files\dotnet\sdk\2.1.400-preview-009063\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.props`

`output2.pp` contains: `F:\corefx\Tools\dotnetcli\sdk\3.0.100-preview1-009019\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.props`

So this is another option, if someone doesn't want to install the SDK into Program Files, add `corefx/Tools/dotnetcli` to their `%PATH%`.

Reply:
Nice Eric. But not sure if we want to point people to internals here. This could and probably will change over time. Though if you think this should be documented I can do that.

Reply:
I think it is likely that we will require pre-release version of .NET SDK for various reasons. The current situation may be pretty normal.

Pointing people to the internal copy of the SDK is going to give them one that is guaranteed to work (it is tested by the CI, etc.). We can have a small .bat/.sh file that adds the internal dotnet to the PATH to make this easy.

If you are going to point people to the global SDK, we have another build configuration to worry about. We would need to make sure that the errors you get from having a wrong version of the global SDK or other configuration problems with the global SDK can be diagnosed. It feels like one of those things that will be always outdated and broken.

Reply:
Are you thinking about setlocals or permanently modifying the path?

Reply:
Permanently modifying the path in given cmd window. Something like this - I have been using a more customized variant of this script  myself:

```
@echo off

set DOTNET_CLI_TELEMETRY_OPTOUT=1
set DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
set DOTNET_MULTILEVEL_LOOKUP=0

set PATH=%~dp0Tools\dotnetcli;%PATH%

title %~dp0
```

Reply:
While I don't mind helper scripts for folks that want them I really want to avoid getting into the old razzle world where they become required.

Reply:
Right that was also my concern. I was glad that we finally could get rid of the Developer Command Prompt requirement. Now introducing another environment seems like a step backwards. I understand Jan's point that in a perfect world we want to use the same tools when building from root and when building individual projects.

I see a few possible options:
a) Recommend to add ~corefx/Tools/dotnetcli/ to the path and put it before the official SDK. That has the disadvantage that you are now using a nightly build of the SDK; not just for corefx but user wide.
b) Add helper scripts (bat & sh) as explained by Jan that sets a few flags and recommend using them.
c) For now try the approach with requiring the 2.1.3 SDK, use that when its released to also build corefx from the root so that it aligns, and monitor how often new SDKs are needed. If regularly depend on nightly builds of the CLI this is not a desirable option.

Reply:
Another idea (not necessarily a good one) is to put a global.json file in the root which indicates the SDK to be used (both for building from root and for individual packages) and install the dotnetcli pre-release version under `%PATH%/dotnet` instead of `~corefx/Tools/dotnetcli/`.

Reply:
+ @stephentoub @ericstj for opinions

Reply:
As someone who moves frequently between Windows and Unix environments, one thing I value (others may not) is consistency across them.  Thus, for example, I like `dotnet msbuild` rather than `msbuild` because the former works correctly everywhere whereas the latter does not.  In that same vein, I'm fine with recommending setting path env vars, as long as whatever we recommend works equally well across OSes (it should).

Reply:
> Another idea (not necessarily a good one) is to put a global.json file in the root which indicates the SDK to be used

We will need a `global.json` file when we move to use `dotnet/arcade`, which requires having a `global.json` that indicates the SDK version to use.

Reply:
I'm very unhappy with how we've conflated the acquisition of an sdk with pinning the sdk version in the ide by using global.json for both. This has decreased my productivity in the sdk repo since we picked up a repo toolset that does this. It doesn't make sense to pin the sdk targets when the ide will use a bunch of other ambient tools (msbuild itself including common targets, compilers, nuget, etc.) 

I prefer to let the IDE use what I have for sdk targets just like it does for everything else.  The failure mode in the IDE is also bugged (will link to project system bug when I'm in the office) when matching version can't be resolved. The pinning also has negative consequences for dogfooding coverage because you end up not trying new sdk targets as they come out even though you do for the rest of the toolset.

We have a lot of improvement to do around this. In the current state, global.json does us more harm than good IMHO. 

I was planning to propose changes to repo toolset / arcade around this.

Reply:
@nguerrera, I believe most of your listed concerns are issues with our official tooling.
* global.json can only 'pin' versions
* The IDE using a mixed set of tool versions (SDK targets vs. MSBuild version)

2 other concerns I have are:
* IDE builds being consistent with command-line builds. I strongly dislike that one can build successfully while the other fails.  Or even worse, that they both build successfully, but give me a different output.
* Build consistency/determinism. If I have the same SHA of a repo, building it on Monday should produce the same output as building it on Tuesday.

I really hope we can address these issues in the shipping product to make it better for both internal .NET Core devs, but even more importantly for our external customers.

Reply:
I agree with you and that's what I meant by "we have a lot of improvement to do".

But I also think that the sdk property in global.json means what it means today and the decision to use this to drive the sdk download in repo toolset is overloading that making in a negative way. This is what I mean by "in current state".

Reply:
I think my dogfooding point is often ignored. There is a lot of value in having our devs actually use the bits they acquire from the dogfooding channel to do actual work. I'm ok with min version over pinning. We don't have this today and even if global.json were min version that doesn't imply that the version official builds download must be the min version to use locally. They should be separate in my very strong opinion.

One might say that we should just update the version used in global.json frequently, but in practice I see it languishing on some arbitrary preview build and then I get angry mail about a break months too late to react easily.

Reply:
> One might say that we should just update the version used in global.json frequently, but in practice I see it languishing on some arbitrary preview build

This is where I should have a pre-made image of the "Maestro" mascot breaking through a wall like the Kool-Aid man:

https://media1.tenor.com/images/0b9cc007367e90f0b2b3439944082f0b/tenor.gif?itemid=4093663

Reply:
That would be cool. Let's do things in the right order. Don't pin unless/until we have a good story for how lkgs are picked up regularly.

Reply:
>I was planning to propose changes to repo toolset / arcade around this.

@nguerrera if you have better ideas now would be a good time to voice them in the arcade repo as we are going to start on-boarding a lot more repos on this plan. I do think auto-updating the SDK more frequently will help but there will always be tension there as we cannot dogfood the latest sdk build due to other requirements such as source-build.

Reply:
Also, I'd still be against mandatory pinning for local builds, but my objection wouldn't be as visceral.

I think official builds should be fully repeatable, but freedom to bring my own tools for unofficial, local work is valuable.

Reply:
> @nguerrera if you have better ideas now would be a good time to voice them in the arcade repo

I will do so.

> but there will always be tension there as we cannot dogfood the latest sdk build due to other requirements such as source-build.

This actually makes my point much more concrete. We have external reasons to ensure that we can build with some particular toolset, but those shouldn't prevent local dogfooding with newer bits. CI can validate that I haven't taken a dependency on something too new without cutting off my ability to dogfood the new stuff.

> But we also need a way where I can tell this build "this is the toolset to use"

Yes, and some one-off gesture shouldn't be needed to dogfood toolsets locally. Source-build will do what it needs to do, but nobody will bother opting in to dogfooding if it doesn't happen naturally/automatically IMHO.